File: ios/FreelanceKar/Views/Sheets/CitySheet.swift
import SwiftUI

struct CitySheetView: View {
    @EnvironmentObject var store: AppStore

    var body: some View {
        let s = store.state
        let rows = Selectors.selectCityRows(s)

        FKSheet(isOpen: s.cityOpen, onClose: { store.closeCity() }, maxHeightFraction: 0.82) {
            VStack(alignment: .leading, spacing: 0) {
                HStack {
                    Text("Where do you need work done?").font(.fkSheetTitle)
                    Spacer()
                    Button { store.closeCity() } label: {
                        Image(systemName: "xmark").font(.system(size: 14)).foregroundStyle(FK.text50)
                    }
                }

                HStack(spacing: 10) {
                    Image(systemName: "magnifyingglass").font(.system(size: 13)).foregroundStyle(FK.text45)
                    TextField("", text: Binding(get: { s.citySearch }, set: { store.setCitySearch($0) }),
                              prompt: Text("Locality or city — e.g. Dharampeth").foregroundStyle(FK.text42))
                        .font(.system(size: 13.5)).foregroundStyle(FK.text)
                }
                .padding(.horizontal, 13).padding(.vertical, 11)
                .background(FK.nested)
                .clipShape(RoundedRectangle(cornerRadius: 13, style: .continuous))
                .overlay(RoundedRectangle(cornerRadius: 13).stroke(FK.line10, lineWidth: 1))
                .padding(.top, 14)

                Button { store.useCurrentLocation() } label: {
                    HStack(spacing: 9) {
                        Text("◉").font(.system(size: 12))
                        Text("Use my current location").font(.system(size: 12.5, weight: .semibold))
                        Spacer()
                    }
                    .foregroundStyle(FK.lime)
                    .padding(.horizontal, 13).padding(.vertical, 11)
                }
                .overlay(RoundedRectangle(cornerRadius: 12).stroke(FK.limeBd35, style: StrokeStyle(lineWidth: 1, dash: [5, 4])))
                .padding(.top, 9)

                VStack(spacing: 7) {
                    ForEach(rows) { r in
                        Button { store.pickPlace(r.city, r.locality) } label: {
                            HStack(spacing: 11) {
                                Text(r.icon).font(.system(size: 12)).foregroundStyle(r.on ? FK.lime : FK.text40)
                                VStack(alignment: .leading, spacing: 2) {
                                    Text(r.name).font(.system(size: 13, weight: .semibold))
                                    Text(r.sub).font(.system(size: 10.5)).foregroundStyle(FK.text45)
                                }
                                Spacer()
                                Text(r.count).font(.fkMonoSm).foregroundStyle(FK.text35)
                            }
                            .padding(.horizontal, 13).padding(.vertical, 12)
                        }
                        .foregroundStyle(FK.text)
                        .background(FK.nested)
                        .clipShape(RoundedRectangle(cornerRadius: 13, style: .continuous))
                        .overlay(RoundedRectangle(cornerRadius: 13).stroke(r.on ? FK.lime : FK.line09, lineWidth: 1))
                    }

                    if rows.isEmpty {
                        VStack(spacing: 14) {
                            Text("No match for “\(s.citySearch)”. FreelanceKar is live in Nagpur, Indore, Jabalpur and Nashik — more cities each month.")
                                .font(.fkBodySm).foregroundStyle(FK.text45).multilineTextAlignment(.center)
                            Button { store.setCitySearch("") } label: {
                                Text("Show all localities").font(.system(size: 12, weight: .semibold)).foregroundStyle(FK.lime)
                                    .padding(.horizontal, 14).padding(.vertical, 10)
                            }
                            .overlay(RoundedRectangle(cornerRadius: 11).stroke(FK.limeBd35, lineWidth: 1))
                        }
                        .padding(.horizontal, 10).padding(.vertical, 26)
                    }
                }
                .padding(.top, 14)
            }
            .padding(.horizontal, 20).padding(.top, 18).padding(.bottom, 22)
        }
    }
}

File: android/app/src/main/java/com/freelancekar/app/ui/sheets/CitySheet.kt
package com.freelancekar.app.ui.sheets

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.SheetState
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.unit.dp
import com.freelancekar.app.data.AppState
import com.freelancekar.app.data.selectCityRows
import com.freelancekar.app.ui.AppViewModel
import com.freelancekar.app.ui.theme.FkColor
import com.freelancekar.app.ui.theme.FkType

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CitySheet(state: AppState, vm: AppViewModel) {
    if (!state.cityOpen) return
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    val rows = selectCityRows(state)

    ModalBottomSheet(
        onDismissRequest = { vm.closeCity() },
        sheetState = sheetState,
        containerColor = FkColor.sheet,
    ) {
        Column(Modifier.fillMaxWidth()) {
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 20.dp).padding(top = 6.dp),
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                Text("Where do you need work done?", style = FkType.sheetTitle, modifier = Modifier.weight(1f))
                Text("✕", style = FkType.body.copy(color = FkColor.text50), modifier = Modifier.clickable { vm.closeCity() })
            }

            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp)
                    .padding(top = 14.dp)
                    .clip(RoundedCornerShape(13.dp))
                    .background(FkColor.nested)
                    .border(1.dp, FkColor.line10, RoundedCornerShape(13.dp))
                    .padding(horizontal = 13.dp, vertical = 2.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("⌕", style = FkType.meta.copy(color = FkColor.text45))
                Spacer(Modifier.width(8.dp))
                TextField(
                    value = state.citySearch,
                    onValueChange = { vm.setCitySearch(it) },
                    placeholder = { Text("Locality or city — e.g. Dharampeth", style = FkType.bodySm.copy(color = FkColor.text42)) },
                    modifier = Modifier.weight(1f),
                    singleLine = true,
                    colors = TextFieldDefaults.colors(
                        focusedContainerColor = FkColor.nested, unfocusedContainerColor = FkColor.nested,
                        focusedIndicatorColor = Color.Transparent, unfocusedIndicatorColor = Color.Transparent,
                    ),
                    textStyle = FkType.bodySm.copy(color = FkColor.text),
                    keyboardOptions = KeyboardOptions(imeAction = ImeAction.Search),
                )
            }

            Row(
                Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 20.dp)
                    .padding(top = 9.dp)
                    .clip(RoundedCornerShape(12.dp))
                    .border(1.dp, FkColor.limeBd35, RoundedCornerShape(12.dp))
                    .clickable { vm.useCurrentLocation() }
                    .padding(horizontal = 13.dp, vertical = 11.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text("◉", style = FkType.meta.copy(color = FkColor.lime))
                Spacer(Modifier.width(9.dp))
                Text("Use my current location", style = FkType.meta.copy(color = FkColor.lime, fontWeight = androidx.compose.ui.text.font.FontWeight.SemiBold))
            }

            Column(Modifier.padding(horizontal = 20.dp).padding(top = 14.dp, bottom = 22.dp), verticalArrangement = Arrangement.spacedBy(7.dp)) {
                rows.forEach { r ->
                    Row(
                        Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(13.dp))
                            .background(FkColor.nested)
                            .border(1.dp, if (r.on) FkColor.lime else FkColor.line09, RoundedCornerShape(13.dp))
                            .clickable { vm.pickPlace(r.city, r.locality) }
                            .padding(horizontal = 13.dp, vertical = 12.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        Text(r.icon, style = FkType.meta.copy(color = if (r.on) FkColor.lime else FkColor.text40))
                        Spacer(Modifier.width(11.dp))
                        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                            Text(r.name, style = FkType.cardTitle)
                            Text(r.sub, style = FkType.metaXs.copy(color = FkColor.text45))
                        }
                        Text(r.count, style = FkType.monoSm.copy(color = FkColor.text35))
                    }
                }

                if (rows.isEmpty()) {
                    Column(Modifier.fillMaxWidth().padding(vertical = 26.dp), horizontalAlignment = Alignment.CenterHorizontally) {
                        Text(
                            "No match for \"${state.citySearch}\". FreelanceKar is live in Nagpur, Indore, Jabalpur and Nashik — more cities each month.",
                            style = FkType.bodySm.copy(color = FkColor.text45),
                            textAlign = androidx.compose.ui.text.style.TextAlign.Center,
                        )
                        Spacer(Modifier.height(14.dp))
                        Row(
                            Modifier
                                .clip(RoundedCornerShape(11.dp))
                                .border(1.dp, FkColor.limeBd35, RoundedCornerShape(11.dp))
                                .clickable { vm.setCitySearch("") }
                                .padding(horizontal = 14.dp, vertical = 10.dp),
                        ) {
                            Text("Show all localities", style = FkType.meta.copy(color = FkColor.lime, fontWeight = androidx.compose.ui.text.font.FontWeight.SemiBold))
                        }
                    }
                }
            }
        }
    }
}
